server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1201,34 +1204,158 @@ void removeEntitlement(Entitlement entitlement,
         sink.queueEvent(event);
     }
 
+
+    /**
+     * Revokes the given set of entitlements.
+     *
+     * @param entsToRevoke entitlements to revoke
+     */
+    @Override
+    @Transactional
+    public void revokeEntitlements(List<Entitlement> entsToRevoke) {
+        if (entsToRevoke.isEmpty()) {
+            return;
+        }
+
+        log.info("Batch revoking entitlements: " + entsToRevoke.size());
+
+        entsToRevoke =  new ArrayList<Entitlement>(entsToRevoke);
+
+        List<Pool> poolsToDelete = poolCurator.listBySourceEntitlements(entsToRevoke);
+
+        for (Pool pool : poolsToDelete) {
+            entsToRevoke.addAll(pool.getEntitlements());
+        }
+
+        for (Entitlement ent : entsToRevoke) {
+            //We need to trigger lazy load of provided products
+            //to have access to those products later in this method.
+            ent.getPool().getProvidedProducts().size();
+            Pool pool = ent.getPool();
+            pool.setConsumed(pool.getConsumed() - ent.getQuantity());
+
+            Consumer consumer = ent.getConsumer();
+            if (consumer.getType().isManifest()) {
+                pool.setExported(pool.getExported() - ent.getQuantity());
+            }
+        }
+
+        log.info("Starting batch delete of pools and entitlements");
+        poolCurator.batchDelete(poolsToDelete);
+        entitlementCurator.batchDelete(entsToRevoke);
+        entitlementCurator.flush();
+        log.info("All deletes flushed successfully");
+
+        List<Entitlement> stackingEntitlements = filterStackingEntitlements(entsToRevoke);
+
+        for (Entitlement ent : stackingEntitlements) {
```

Comment:
It seems like we are looping through entsToRevoke (or a subset thereof) three times: the first loop to change the pool consumed values, the second to search for stacking entitlements, and the the third to loop through the list of stacking entitlements we just received.  Can we consolidate?

```
for (Entitlement e : entsToRevoke)
  //get pool
  //adjust pool quantity
  if (isStackingEntitlement(e)) {
    //do stacking stuff or append to stackingEntitlements list
  }
}

// do stacking stuff here if not possible in previous loop
```